distributed-counter/src/main/java/org/infinispan/counter/CounterManager.java

```diff
@@ -0,0 +1,131 @@
+      this.cacheManager = cacheManager;
+      this.counters = new ConcurrentHashMap<>();
+   }
+
+   /**
+    * @return the {@link CounterManager} associated to the {@link EmbeddedCacheManager}.
+    */
+   public static CounterManager getCounterManager(EmbeddedCacheManager cacheManager) {
+      return requireNonNull(cacheManager, "EmbeddedCacheManager can't be null.")
+            .getGlobalComponentRegistry()
+            .getComponent(CounterManager.class);
+   }
+
+   /**
+    * Register a {@link CounterManager} instance to the {@link EmbeddedCacheManager}.
+    * <p>
+    * This method is invoked automatically when a {@link EmbeddedCacheManager} starts.
+    */
+   public static void registerCounterManager(EmbeddedCacheManager cacheManager) {
+      GlobalComponentRegistry registry = requireNonNull(cacheManager).getGlobalComponentRegistry();
+      //just o avoid creating multiple counter manager.
+      //noinspection SynchronizationOnLocalVariableOrMethodParameter
+      synchronized (registry) {
+         CounterManager counterManager = registry.getComponent(CounterManager.class);
+         if (counterManager == null) {
+            counterManager = new CounterManager(cacheManager);
+            registry.registerComponent(counterManager, CounterManager.class);
+         }
+      }
+   }
+
+   /**
+    * It returns the {@link Counter} with that specific name.
+    * <p>
+    * If the {@link Counter} does not exists, it is created. Only the first invocation will take in consideration the
+    * initial value.
+    *
+    * @param name         the counter name.
+    * @param initialValue the counter initial value.
+    * @return the {@link Counter} instance.
+    */
+   public Counter getOrCreateCounter(String name, long initialValue) {
+      return counters.computeIfAbsent(name, counterName -> {
+         InfinispanCounter counter = new InfinispanCounter(counterName, cacheManager.getCache(COUNTER_CACHE_NAME));
+         counter.init(initialValue);
+         return counter;
+      });
+   }
+
+   /**
+    * It returns the {@link Counter} with that specific name.
+    * <p>
+    * If the {@link Counter} does not exists, it is created. Only the first invocation will take in consideration the
+    * initial value, min and max thresholds.
+    * <p>
+    * The initial value must be between the min and max value.
+    *
+    * @param name         the counter name.
+    * @param initialValue the counter initial value.
+    * @return the {@link Counter} instance.
+    * @throws IllegalArgumentException if the min, max or initial value are not valid.
+    */
+   public Counter getOrCreateCounter(String name, long initialValue, long minThreshold, long maxThreshold) {
+      if (minThreshold >= initialValue || initialValue >= maxThreshold) {
+         throw new IllegalArgumentException("Initial value must be between the min and max threshold.");
+      }
+      return counters.computeIfAbsent(name, counterName -> {
```

Comment:
So if there's another counter with different thresholds, and possibly even weak one, defined, you return that instead without any warning? I would rather check that the configuration (weakness + thresholds) match to the existing one, and throw exception if there's mismatch.